libsigner/src/events.rs

```diff
@@ -0,0 +1,268 @@
+                info!("Event receiver stopped");
+                break;
+            }
+            let next_event = match self.next_event() {
+                Ok(event) => event,
+                Err(EventError::UnrecognizedEvent(..)) => {
+                    // got an event that we don't care about (not a problem)
+                    continue;
+                }
+                Err(EventError::Terminated) => {
+                    // we're done
+                    info!("Caught termination signal");
+                    break;
+                }
+                Err(e) => {
+                    warn!("Failed to receive next event: {:?}", &e);
+                    continue;
+                }
+            };
+            if !self.forward_event(next_event) {
+                info!("Failed to forward event");
+                break;
+            }
+        }
+        info!("Event receiver main loop exit");
+    }
+}
+
+pub struct StackerDBEventReceiver {
+    /// contracts we're listening for
+    pub stackerdb_contract_ids: Vec<QualifiedContractIdentifier>,
+    /// Address we bind to
+    local_addr: Option<SocketAddr>,
+    /// server socket that listens for HTTP POSTs from the node
+    sock: Option<TcpListener>,
+    /// channel into which to write newly-discovered data
+    out_channels: Vec<Sender<StackerDBChunksEvent>>,
+    /// inter-thread stop variable -- if set to true, then the `main_loop` will exit
+    stop_signal: Arc<AtomicBool>,
+}
+
+impl StackerDBEventReceiver {
+    /// Make a new StackerDB event receiver, and return both the receiver and the read end of a
+    /// channel into which node-received data can be obtained.
+    pub fn new(contract_ids: Vec<QualifiedContractIdentifier>) -> StackerDBEventReceiver {
+        let stackerdb_receiver = StackerDBEventReceiver {
+            stackerdb_contract_ids: contract_ids,
+            sock: None,
+            local_addr: None,
+            out_channels: vec![],
+            stop_signal: Arc::new(AtomicBool::new(false)),
+        };
+        stackerdb_receiver
+    }
+
+    /// Do something with the socket
+    pub fn with_socket<F, R>(&mut self, todo: F) -> Result<R, EventError>
+    where
+        F: FnOnce(&mut StackerDBEventReceiver, &mut TcpListener) -> R,
+    {
+        let mut sock = if let Some(s) = self.sock.take() {
+            s
+        } else {
+            return Err(EventError::NotBound);
+        };
+
+        let res = todo(self, &mut sock);
+
+        self.sock = Some(sock);
+        Ok(res)
+    }
+}
+
+/// Stop signaler implementation
+pub struct StackerDBStopSignaler {
+    stop_signal: Arc<AtomicBool>,
+    local_addr: SocketAddr,
+}
+
+impl StackerDBStopSignaler {
+    pub fn new(sig: Arc<AtomicBool>, local_addr: SocketAddr) -> StackerDBStopSignaler {
+        StackerDBStopSignaler {
+            stop_signal: sig,
+            local_addr,
+        }
+    }
+}
+
+impl EventStopSignaler for StackerDBStopSignaler {
+    fn send(&mut self) {
```

Comment:
Nit: I sugesst using a more explicit name here.
```suggestion
    fn send_stop_signal(&mut self) {
```

Reply:
Not necessarily for all structs implementing the trait, but sure - for the StackerDBStopSignaler I wouldn't imagine anything else being sent through this method.